internal/mempool/mempool_test.go

```diff
@@ -2285,3 +2313,239 @@ func TestRemoveDoubleSpends(t *testing.T) {
+		return true, nil
+	}
+	harness.txPool.cfg.Policy.MaxTxVersion = 3
+	harness.chain.SetStandardVerifyFlags(BaseStandardVerifyFlags | txscript.ScriptVerifyTreasury)
+	nextHeight := net.StakeValidationHeight
+	harness.chain.SetHeight(nextHeight - 1)
+
+	// Useful constants.
+	tvi := net.TreasuryVoteInterval
+	mul := net.TreasuryVoteIntervalMultiplier
+
+	// Helper to assert that the TSpendHashes() function returns the
+	// correct tspends.
+	assertTSpendHashes := func(tspends []*dcrutil.Tx) {
+		t.Helper()
+		tspendHashes := harness.txPool.TSpendHashes()
+		tspendHashesMap := make(map[chainhash.Hash]struct{}, len(tspendHashes))
+		for i := 0; i < len(tspendHashes); i++ {
+			tspendHashesMap[tspendHashes[i]] = struct{}{}
+		}
+		for i := 0; i < len(tspends); i++ {
+			h := tspends[i].MsgTx().TxHash()
+			if _, ok := tspendHashesMap[h]; !ok {
+				t.Fatalf("added tspend %d was not returned in TSpendHashes()", i)
+			}
+			delete(tspendHashesMap, h)
+		}
+		if len(tspendHashesMap) > 0 {
+			t.Fatalf("TSpendHashes() returned extraneous data: %v", tspendHashesMap)
+		}
+	}
+
+	// Helper that adds and asserts the given tspend was added to the
+	// mempool.
+	acceptTSpend := func(tx *dcrutil.Tx) {
+		t.Helper()
+		_, err = harness.txPool.ProcessTransaction(tx, true,
+			false, true, 0)
+		if err != nil {
+			t.Fatalf("ProcessTransaction: failed to accept valid tspend %v", err)
+		}
+		testPoolMembership(tc, tx, false, true)
+	}
+
+	// Helper that attempts to add and asserts the given tspend is rejected
+	// with the given error code.
+	rejectTSpend := func(tx *dcrutil.Tx, errCode ErrorCode) {
+		t.Helper()
+		_, err = harness.txPool.ProcessTransaction(tx, true,
+			false, true, 0)
+		if !IsErrorCode(err, errCode) {
+			t.Fatalf("Unexpected error while processing rejected tspend. "+
+				"want=%v got=%#v", errCode, err)
+		}
+		testPoolMembership(tc, tx, false, false)
+	}
+
+	// Helper that attempts to add and asserts the given tspend is rejected
+	// with the given blockchain error code.
+	//
+	// TODO: unify with the above after mempool errors correctly handle
+	// errors.Is/As.
+	rejectTSpendChainError := func(tx *dcrutil.Tx, errCode blockchain.ErrorCode) {
+		t.Helper()
+		_, err = harness.txPool.ProcessTransaction(tx, true,
+			false, true, 0)
+		rerr, ok := err.(RuleError)
+		if !ok {
+			t.Fatalf("Returned error is not a rule error: %#v", err)
+		}
+		if !blockchain.IsErrorCode(rerr.Err, errCode) {
+			t.Fatalf("Unexpected error while processing rejected tspend. "+
+				"want=%v got=%#v", errCode, err)
+		}
+		testPoolMembership(tc, tx, false, false)
+	}
+
+	expiry := standalone.CalculateTSpendExpiry(nextHeight, tvi, mul)
+	tspendAmount := int64(1e8)
+	tspendFee := int64(2550)
+	valueIn := tspendAmount + tspendFee
+	piKey, err := hex.DecodeString("62deae1ab2b1ebd96a28c80e870aee325bed359e83d8db2464ef999e616a9eef")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create a few valid tspends that can enter the mempool.
+	nbTSpends := 8
```

Comment:
Doesn't matter too much since it's a test, but this could be a const since it doesn't change.  The same is true in various other cases.

I'd also define `maxTSpends` just before this line and define this one in terms of it +1 to ensure the below code can't easily panic if the max value is changed.